File: app/javascript/channels/chats_channel.js
import consumer from "channels/consumer"

const channel = consumer.subscriptions.create({channel: "ChatsChannel", chapter_id: 20}, {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    console.log(data)
  }
});

window.channel = channel;

File: app/channels/chats_channel.rb
class ChatsChannel < ApplicationCable::Channel
  Message = Struct.new(:chat_message, :user)

  def subscribed
    @chat = Chat.create_or_find_by!(chapter_id: params["chapter_id"], user_id: current_user.id)
    stream_for @chat
    logger.info("stream started for user(#{current_user.id}) for chat(#{@chat.id})")
  end

  def unsubscribed
    stop_stream_for(@chat)
    logger.info("stream stopped for user(#{@chat.user_id}) for chat(#{@chat.id})")
  end

  def receive(data)
    logger.info("received message in channel: #{data["message"]}")
    chat_message = @chat.chat_messages.create!(user: current_user, message: data["message"])
    RespondToChatJob.perform_later(chat_message)
  end
end
